### ISSUE TITLE
Variable re-declaration can cause duplicate allocas

### ISSUE THREAD

**Issue by GlowingScrewdriver** [chsasank/llama.lisp#10]
As of now, an `alloca` instruction is added for every single variable assignment, instead of the first assignment. This is a bug.
The effect is that each instance of a variable is mapped to a differently numbered `alloca` pointer, causing the mem2reg pass to insert phi `undef` values in the resulting phi instructions.

Consider the following Bril program:
```
@main: int {
    count: int = const 3;
    one: int = const 1;
.loop:
    count: int = sub count one;
    end: bool = eq count one;
    br end .loop .end;

.end:
    ret one;
}
```
When this is compiled to LLVM IR, by `llvm.py`, we see something like this in the `alloca` basic block:
```
alloca_eQdrYwCnXP:
  %"count" = alloca i32
  %"one" = alloca i32
  %"count.1" = alloca i32
```

And finally, when passed through mem2reg,
```
loop:                                             ; preds = %loop, %entry_osrxUvDtzg
  %count.1.0 = phi i32 [ undef, %entry_osrxUvDtzg ], [ %count.2, %loop ]
  %count.2 = sub i32 %count.1.0, 1
```

**Comment by GlowingScrewdriver** [comment]
> As of now, an `alloca` instruction is added for every single variable assignment, instead of the first assignment. This is a bug.

The reason this happens is the [following line in `llvm.py`](https://github.com/chsasank/llama.lisp/blob/d31b7b7ed5617f1c86fe81dd30fcef7ddb42783a/src/backend/llvm.py#L213):
``` python3
if (name not in self.func_alloca_symtab) or (self.func_alloca_symtab[name].type != typ):
```
The new variable's type is being compared to a _pointer_ to the existing variable's type. It should have been:
``` python3
if (name not in self.func_alloca_symtab) or (self.func_alloca_symtab[name].type != typ.as_pointer()):
```

**Comment by GlowingScrewdriver** [comment]
Do we even want to support variable re-declaration with different types?
* C doesn't allow it
* Bril allows it

We need to answer the question because
* Bril has a test for non-linear control flow, wherein some variable references are physically placed before the variable assignments, but the assignments come first in the CFG.
* So to allow that, we have to scan over the entire function first, creating `alloca`s for all variables, and then move on to process the instructions.
* In that case we need to have a symbol table that holds all combinations of {variable name, type}.
* This also means that for variable access, both variable name and type are required (currently, only name is required)

**Comment by GlowingScrewdriver** [comment]
> Bril has a [test for non-linear control flow](https://github.com/sampsyo/bril/blob/main/test/interp/core/non_linear_control_flow.bril), wherein some variable references are physically placed before the variable assignments, but the assignments come first in the CFG.
We are not going to try passing the test for non-linear control flow. C Requires all declarations to be before the references, so we'll follow the same.

**Comment by chsasank** [comment]
Fixed by #11
